tests/unittests/CoreGraphics.drawing/CGContextDrawing_ImageMaskTests.cpp

```diff
@@ -0,0 +1,314 @@
+        ClippingShape shape = ::testing::get<0>(GetParam());
+        ClippingType type = ::testing::get<1>(GetParam());
+        const char* shapeName = "unk";
+        const char* typeName;
+        switch (shape) {
+            case ClippingShapeDiamond:
+                shapeName = "di";
+                break;
+            case ClippingShapeSquare:
+                shapeName = "sq";
+                break;
+            case ClippingShapeRectangle:
+                shapeName = "rct";
+                break;
+        }
+        switch (type) {
+            case ClippingTypeMask:
+                typeName = "mask";
+                break;
+            case ClippingTypeAlpha:
+                typeName = "alpha";
+                break;
+        }
+        return CFStringCreateWithFormat(nullptr,
+                                        nullptr,
+                                        CFSTR("TestImage.CGContextClipping.%s.%s.%s.png"),
+                                        test_info->name(),
+                                        shapeName,
+                                        typeName);
+    }
+
+    CGImageRef GetClippingImage(ClippingShape shape, ClippingType type) {
+        static ClippingType s_clippingTypes[]{
+            ClippingTypeAlpha, ClippingTypeMask,
+        };
+        static std::unordered_map<unsigned int, woc::unique_cf<CGImageRef>> s_clippingImages = []() -> decltype(s_clippingImages) {
+            decltype(s_clippingImages) clippingImages;
+            for (auto& tpl : s_maskGenerators) {
+                for (auto& newType : s_clippingTypes) {
+                    ClippingShape newShape = std::get<0>(tpl);
+                    woc::unique_cf<CGImageRef> clipImage{ __CreateClipImage(std::get<1>(tpl), newType, std::get<2>(tpl)) };
+                    clippingImages.emplace(newShape | newType, std::move(clipImage));
+                }
+            }
+            return std::move(clippingImages);
+        }();
+        return s_clippingImages.at(shape | type).get();
+    }
+
+    void _FillContextWithColorAndLines() {
+        CGContextRef context = GetDrawingContext();
+        CGRect bounds = GetDrawingBounds();
+        CGContextSetRGBStrokeColor(context, 1.0, 0.0, 0.0, 1.0);
+        CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1.0);
+
+        CGContextFillRect(context, bounds);
+
+        for (CGFloat i = 0.5f; i < bounds.size.width; i += 10.0f) {
+            CGContextMoveToPoint(context, i, 0);
+            CGContextAddLineToPoint(context, i, bounds.size.height);
+            CGContextStrokePath(context);
+        }
+    }
+};
+
+std::vector<std::tuple<ClippingShape, CGSize, std::function<void(CGContextRef, CGSize, ClippingType)>>> CGContextClipping::s_maskGenerators{
+    { ClippingShapeSquare, CGSize{ 128, 128 }, __MakeSquareMaskGenerator(CGAffineTransformIdentity) },
+    { ClippingShapeDiamond, CGSize{ 128, 128 }, __MakeSquareMaskGenerator(CGAffineTransformMakeRotation(45.f * M_PI / 180.f)) },
```

Comment:
Nah: That's not the thing on test here.

---
In reply to: [92094286](https://github.com/Microsoft/WinObjC/pull/1528#discussion_r92094286) [](ancestors = 92094286)